### PR TITLE
[Unification] Make get_clones a common utility

### PR DESCRIPTION
**Pull request by langong347**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #205
* __->__ #213
* #204
* #176

Summary:
- Move `get_clones` to `util/` as a general utility for encoder/decoders

Test Plan:
```
$ python -m pytest test/models/test_gpt.py test/models/test_mdetr.py -vv
============================================================ test session starts ============================================================
platform darwin -- Python 3.8.13, pytest-7.1.2, pluggy-1.0.0 -- /Users/langong/local/miniconda3/envs/torchmm/bin/python
cachedir: .pytest_cache
rootdir: /Users/langong/gpt_attention, configfile: pyproject.toml
plugins: mock-3.8.2, cov-3.0.0
collected 17 items

test/models/test_gpt.py::TestMultimodalGPT::test_bad_input PASSED                                                                     [  5%]
test/models/test_gpt.py::TestMultimodalGPT::test_forward_in_modality PASSED                                                           [ 11%]
test/models/test_gpt.py::TestMultimodalGPT::test_forward_out_modality PASSED                                                          [ 17%]
test/models/test_gpt.py::TestMultimodalGPT::test_forward_two_modality PASSED                                                          [ 23%]
test/models/test_gpt.py::TestMultimodalGPT::test_encode PASSED                                                                        [ 29%]
test/models/test_gpt.py::TestMultimodalGPT::test_bad_pos_ids PASSED                                                                   [ 35%]
test/models/test_gpt.py::TestMultimodalGPT::test_optional_pos_ids PASSED                                                              [ 41%]
test/models/test_gpt.py::TestTransformerDecoder::test_forward PASSED                                                                  [ 47%]
test/models/test_gpt.py::TestTransformerDecoder::test_forward_additional_output PASSED                                                [ 52%]
test/models/test_gpt.py::TestTransformerDecoderLayer::test_forward PASSED                                                             [ 58%]
test/models/test_gpt.py::TestTransformerDecoderLayer::test_forward_masked PASSED                                                      [ 64%]
test/models/test_gpt.py::TestTransformerDecoderLayer::test_forward_additional_output PASSED                                           [ 70%]
test/models/test_gpt.py::test_sigmoid_linear_unit PASSED                                                                              [ 76%]
test/models/test_gpt.py::test_right_shift PASSED                                                                                      [ 82%]
test/models/test_mdetr.py::TestMDETR::test_transformer_encoder PASSED                                                                 [ 88%]
test/models/test_mdetr.py::TestMDETR::test_transformer_decoder PASSED                                                                 [ 94%]
test/models/test_mdetr.py::TestMDETR::test_full_mdetr_model PASSED                                                                    [100%]

====================== 17 passed in 3.92s =========================
```

Differential Revision: [D38283964](https://our.internmc.facebook.com/intern/diff/D38283964)

### CODE REVIEW

File: torchmultimodal/utils/common.py [hunk 1]
```diff
@@ -216,3 +217,7 @@ def custom_forward(*inputs: Any) -> Callable:
             return fn(cls, *inputs, **kwargs)
 
     return inner
+
+
+def get_clones(module: nn.Module, n: int) -> nn.Module:
+    return nn.ModuleList([deepcopy(module) for i in range(n)])
```
Comment (mthrok) [inline torchmultimodal/utils/common.py:223]: Does PyTorch officially state that `deepcopy`-ing `nn.Module` keep working for future?

**Review by RdoubleA — APPROVED** [review]

File: torchmultimodal/utils/common.py (hunk 1 above)
Comment (langong347) [reply torchmultimodal/utils/common.py:223]: Thanks for the comment. I haven't noticed this before. Would you mind sharing some details?
For context, `deepcopy` is used by PT core's own implementation to stack up transformer layers [here](https://pytorch.org/docs/stable/_modules/torch/nn/modules/transformer.html#TransformerDecoder).

Comment (mthrok) [reply torchmultimodal/utils/common.py:223]: I cannot find the `__deepcopy__` protocol implemented on [torch.nn.Module](https://github.com/pytorch/pytorch/blob/master/torch/nn/modules/module.py). The multiple Q&A just says it works. So I was wondering if it's tested and ensured in PyTorch core.

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38283964).

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38283964).

**Comment by langong347** [comment]
@langong347 has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D38283964).

**Pull request closed without merging**